dnssec-backend.go

```diff
@@ -0,0 +1,313 @@
+	return len(z.Dnskey.RrSet) > 0
+}
+
+func (z SignedZone) lookupPubKey(keyTag uint16) *dns.DNSKEY {
+	return z.PubKeyLookup[keyTag]
+}
+
+func (z SignedZone) addPubKey(k *dns.DNSKEY) {
+	z.PubKeyLookup[k.KeyTag()] = k
+}
+
+func removeRRSIGs(response *dns.Msg) *dns.Msg {
+	var filteredAnswers []dns.RR
+	filteredResponse := response.Copy()
+	for _, rr := range filteredResponse.Answer {
+		if _, isRRSIG := rr.(*dns.RRSIG); !isRRSIG {
+			filteredAnswers = append(filteredAnswers, rr)
+		}
+	}
+
+	filteredResponse.Answer = filteredAnswers
+	return filteredResponse
+}
+
+func setDNSSECdo(q *dns.Msg) *dns.Msg {
+	if q.IsEdns0() == nil {
+		q.SetEdns0(1024, true)
+	}
+	q.IsEdns0().SetDo()
+	return q
+}
+
+func newQuery(qname string, qtype uint16) *dns.Msg {
+	dnsMessage := &dns.Msg{
+		MsgHdr: dns.MsgHdr{
+			RecursionDesired: true,
+		},
+	}
+	dnsMessage.SetEdns0(4096, true)
+	dnsMessage.SetQuestion(qname, qtype)
+	return dnsMessage
+}
+
+func getRRset(qname string, qtype uint16, resolver Resolver, ci ClientInfo) (*RRSet, error) {
+	q := newQuery(qname, qtype)
+	r, err := doQuery(q, resolver, ci)
+	if err != nil || r == nil {
+		return nil, err
+	}
+	return extractRRset(r)
+}
+
+func extractRRset(r *dns.Msg) (*RRSet, error) {
+	result := &RRSet{RrSet: make([]dns.RR, 0)}
+	if r.Answer == nil {
+		return result, nil
+	}
+
+	result.RrSet = make([]dns.RR, 0, len(r.Answer))
+	for _, rr := range r.Answer {
+		switch t := rr.(type) {
+		case *dns.RRSIG:
+			result.RrSig = t
+		case *dns.DNSKEY, *dns.DS, *dns.A, *dns.AAAA:
+			result.RrSet = append(result.RrSet, rr)
+		}
+	}
+	return result, nil
+}
+
+func doQuery(q *dns.Msg, resolver Resolver, ci ClientInfo) (*dns.Msg, error) {
+	r, err := resolver.Resolve(q, ci)
+	if err != nil {
+		return nil, err
+	}
+	if r.Rcode == dns.RcodeNameError {
+		Log.Info("no such domain", "info", qName(r))
+		return nil, ErrNoResult
+	}
+	if r == nil || r.Rcode == dns.RcodeSuccess {
+		return r, err
+	}
+	return nil, ErrInvalidRRsig
+}
+
+func queryDelegation(domainName string, resolver Resolver, ci ClientInfo) (*SignedZone, error) {
+	signedZone := &SignedZone{
+		Zone:   domainName,
+		Ds:     &RRSet{},
+		Dnskey: &RRSet{},
+	}
+	signedZone.PubKeyLookup = make(map[uint16]*dns.DNSKEY)
+
+	var g errgroup.Group
+	g.Go(func() error {
+		var err error
+		signedZone.Dnskey, err = getRRset(domainName, dns.TypeDNSKEY, resolver, ci)
+		if err != nil {
+			return err
+		}
+
+		for _, rr := range signedZone.Dnskey.RrSet {
+			defer func() {
+				if r := recover(); r != nil {
+					Log.Warn("panic occurred",
+						slog.String("domainName", domainName),
+						slog.Any("panic", r),
+					)
+				}
+			}()
+			signedZone.addPubKey(rr.(*dns.DNSKEY))
+		}
+		return nil
+	})
+
+	g.Go(func() error {
+		var err error
+		signedZone.Ds, err = getRRset(domainName, dns.TypeDS, resolver, ci)
+		if err != nil {
+			Log.Error("DS query failed",
+				slog.String("domainName", domainName),
+				"error", err,
+			)
+
+			return err
```

Comment:
Rather than logging from within, might be better to extend the error and pass it back to the caller to handle. Something like

```go
return fmt.Errorf("DS query failed for %q: %w", domainName, err)
```